autoagora_agents/controller.py

```diff
@@ -0,0 +1,68 @@
+# Copyright 2022-, Semiotic AI, Inc.
+# SPDX-License-Identifier: Apache-2.0
+
+from typing import Any
+
+import numpy as np
+import torch
+
+from autoagora_agents.algorithm import Algorithm, algorithmgroupfactory
+
+
+class Controller:
+    """Holds all algorithms and routes information to each.
+
+    Keyword Arguments:
+        agents (list[dict[str, Any]]): A list of the configs for each agent
+            group.
+        seed (int): The seed for torch.
+
+    Attributes:
+        groups (dict[str, Algorithm]): A dictionary mapping agent groups to algorithms.
+    """
+
+    def __init__(self, *, agents: list[dict[str, Any]], seed: int) -> None:
+        self.groups = {a["group"]: algorithmgroupfactory(**a) for a in agents}
+        torch.manual_seed(seed)
+
+    def __call__(
+        self,
+        *,
+        observations: dict[str, np.ndarray],
+        actions: dict[str, np.ndarray],
+        rewards: dict[str, float],
+        dones: dict[str, bool]
+    ) -> dict[str, np.ndarray]:
+        """Call each algorithm.
+
+        Keyword Arguments:
+            observations (dict[str, np.ndarray]): The observations of each agent.
+            actions (dict[str, np.ndarray]): The action of each agent.
+            rewards (dict[str, float]): The reward received by each agent.
+            dones (dict[str, bool]): Whether each agent is done.
```

Comment:
eheheh, terminal states, you mean? :)

Reply:
The RL terminology is `dones` haha. And it's a boolean. Terminal states would be the actual state they terminated in, no?